4단계 - 자동차 경주(우승자)

h2jinee opened the pull request:
## 📌 Summary

- 3단계 미션에서 피드백 주셨던 `getPositionDisplay()` 메서드를 View가 담당하도록 리팩토링했습니다.
- 4단계 자동차 경주 미션을 진행했습니다. 어떻게 손을 댈지 막막해 라이브 강의에서 보여주셨던 것처럼 굴러가게끔 우선 구현했습니다 🥹
    - `CarTest`를 작성하던 시점에 테스트 코드를 어떻게 구현해야 할지 조차 감이 오지 않아서 객체지향의 사실과 오해 책을 같이 보면서 진행했습니다. (아직 2장까지만 읽었습니다..)
    - 테스트 코드를 고민을 해도 어떻게 작성해야 할지 모르겠어서 실제 돌아가는 코드부터 만들어 테스트 해보았습니다.
    - 아직 테스트를 제대로 작성하지 못한 부분들도 있고 리팩토링 해야 하는 부분들도 많은데, 도저히 감이 오지 않아 리뷰 받으면서 개선하고 싶습니다.

### **🤔 막혔던 부분 및 질문**

- 단위 테스트를 작성하던 중 '각 자동차의 이름은 중복될 수 없다.' 테스트부터 기존에 작성해 놓은 코드가 많은데 여기서 누가 중복 검증에 대한 책임을 담당해야 하는지?에 대해서 고민이 많았습니다.
- **질문** : 3단계에서 작성한 테스트 코드들이 4단계에 코드를 수정하면서 완전히 실패하게 된다면, 테스트 코드를 잘못 작성한 건가요?
- RacingGame에서 `addCars()` 관련이 제일 고민이 많았습니다.. `names`가 파라미터로 들어오는 경우를 따로 고려해서 또 새로운 것을 만드는 게 스스로 이상하다고 생각이 들었는데 어떻게 개선해야 되지?에 대해서 답이 떠오르지 않았습니다.
- RacingGame에 `addCars()`가 들어가는 게 맞나?에 대한 고민이 정말 정말 많았습니다.. 마찬가지로 `Car`에 이 메서드가 들어가는 게 맞나?에 대한 고민이 많았습니다.

h2jinee committed:
refactor: "-"를 View가 담당하도록 변경

h2jinee committed:
test: 4단계 - 자동차 경주(우승자) 테스트 코드 추가

h2jinee committed:
refactor: 자동차 등록 시 중복 검증 수정

h2jinee committed:
refactor: 프로젝트 구조 수정

h2jinee committed:
refactor: 테스트 데이터 경계값으로 수정

h2jinee committed:
feat: 자동차 경주(우승자) 구현

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,71 @@
+
+    public Car() {
+        this.name = "";
+        this.position = 0;
```
```suggestion
        this("");
```
부 생성자가 주 생성자를 호출하는 방식으로 구현하는 습관 들일 것을 추천
그런데 자동차 이름이 없으면 아예 Car가 생성되면 안될 것 같은데 이 생성자가 필요할까?

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,71 @@
+        return position;
+    }
+
+    public String move(int number) {
```
👍

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,71 @@
+            return getPositionDisplay(this);
+        }
+
+        return getPositionDisplayWithNames(this);
```
이렇게 구현하면 move 메서드는 두 가지 책임을 가지게 된다.
출력하는 값을 생성하는 부분은 View가 담당하도록 구현하면 어떨까?

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,57 @@
+
+    public RacingGame(String names) {
+        splitCarName(names);
+    }
```
```suggestion
    public RacingGame(String names) {
        this(createCars(names));
    }
    
    public RacingGame(List<Car> cars) {
        this.cars = cars;
    }
```
필드 타입과 같은 주 생성자를 추가하고, 부 생성자가 주 생성자를 호출하는 방식으로 구현하는 습관을 들인다.

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,57 @@
+        for (int i = 0; i < number; i++) {
+            cars.add(new Car());
+        }
+    }
```
이름을 가지지 않는 생성자와 메서드가 왜 필요할까?
이 코드를 필요없는 코드이지 않을까?

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,57 @@
+        for (String s : split) {
+            cars.add(new Car(s.trim()));
+        }
+    }
```
```suggestion
    private static List<Car> createCars(String names) {
        String[] split = names.split(",");

        List<Car> cars = new ArrayList<>();
        for (String s : split) {
            cars.add(new Car(s.trim()));
        }
        return cars;
    }
```
위와 같이 구현하는 것은 어떨까?

javajigi commented on src/main/java/racinggame/view/OutputView.java:
```diff
@@ -0,0 +1,52 @@
+            if (car.getPosition() == maxPosition) {
+                winners.add(car.getName());
+            }
+        }
```
우승자를 구하는 로직은 핵심 도메인 로직으로 보인다.
우승자를 구하기 위한 자동차 목록을 가지고 있는 RacingGame이 담당하면 어떨까?


```suggestion
        List<Car> winners = racingGame.findWinners();
```

javajigi commented on src/test/java/racinggame/domain/CarTest.java:
```diff
@@ -0,0 +1,47 @@
+    @DisplayName("각 자동차의 이름은 중복될 수 없다")
+    void carNamesCannotBeDuplicated() {
+
+    }
```
중복 체크를 한다면 names를 split한 후 List<Car>를 생성하는 곳에서 해야하지 않을까?
이럴 경우 메서드 복잡도가 높아진다면 names에서 List<Car> 생성하는 역할을 별도의 객체로 분리하고 테스트하는 것도 좋은 접근 방법이 될 수 있다.
예를 들어 CarFactory.createCars(String names) 와 같이

javajigi commented on src/test/java/racinggame/domain/WinnerTest.java:
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class WinnerTest {
```
우승자를 구하는 로직이 도메인 객체가 아닌 View에 있다보니 테스트하기 어려움
상태 값을 가지는 RacingGame과 같은 객체가 우승자를 구하도록 접근해 본다.

javajigi reviewed (changes requested):
4단계 진행함에 있어 어려움이 많았던 것으로 알고 있는데요.
그래도 포기하지 않고 끝끝내 구현한 점 칭찬하고 싶네요. 👍 
지금까지 객체를 분리하고, 객체에 책임을 부여하는 연습이 되어 있지 않다보니 어려움이 많았을 것 같아요. 괜찮습니다.
지금부터 연습하면 좋아질테니까요.
수업 중 다뤘던 추가적인 객체 분리는 다음 피드백에서 진행해 보고요.
지금 단계는 피드백한 내용을 반영하는 위주로 접근해 보면 좋겠습니다.
피드백 반영했다 생각하면 바로 리뷰 요청 주세요.

Q1. 3단계에서 작성한 테스트 코드들이 4단계에 코드를 수정하면서 완전히 실패하게 된다면, 테스트 코드를 잘못 작성한 건가요?
A1. 이 같은 상황이 발생한 이유는 테스트 코드보다 프로덕션 코드를 먼저 수정했기 때문입니다.
TDD에서는 무엇인가 요구사항이 추가되거나 변경되었을 때 프로덕션 코드보다 테스트 코드를 먼저 수정한 후 프로덕션 코드를 수정하면 이런 문제가 발생하지 않는데요. 보통 프로덕션 코드를 먼저 수정하기 때문에 발생한 이슈일 겁니다.

h2jinee committed:
refactor: Car 클래스 기본 생성자 제거 및 테스트 데이터 수정

- Car 클래스에서 기본 생성자를 제거하고, 이름을 빈 문자열로 설정하는 생성자를 사용하도록 변경
- RacingGame 클래스에서 자동차 추가 시 빈 문자열을 전달하도록 수정
- RacingTest 클래스에서 테스트를 위한 자동차 이름을 "test"로 설정

h2jinee committed:
refactor: RacingGame 클래스 생성자 수정 및 Car 클래스 메서드 정리

- RacingGame 클래스의 생성자를 수정하여 자동차 목록을 직접 받도록 변경
- Car 클래스의 getPositionDisplay 메서드에서 이름이 비어있을 때의 처리 로직 수정
- OutputView 클래스에서 getPositionDisplayWithNames 메서드를 제거하고, 이름이 비어있을 때의 처리 로직 추가

h2jinee commented:
@javajigi 포비님 `carNamesCannotBeDuplicated()` 테스트와 우승자 관련해서 피드백 주신 것은 아직 리팩토링 중인데 시간이 조금 더 걸릴 것 같아 우선 리팩토링 완료한 부분까지 리뷰 요청드립니다. 아직 리팩토링을 하지 못한 부분들도 금일내로 수정해서 요청드리겠습니다!

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.List;
+
+public class RacingGame {
+    private static List<Car> cars = new ArrayList<>();
```
```suggestion
    private final List<Car> cars;
```
상태 값을 가지는 인스턴스 변수는 static으로 구현할 경우 멀티 쓰레드 환경에서 문제가 발생할 수 있음.
특별한 이유가 없는 한 위와 같이 인스턴스 변수로 구현할 것을 추천

이번 기회에 클래스와 인스턴스의 차이점에 대해 추가 학습하는 시간으로 만들어볼 것을 추천

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,54 @@
+
+        for (String s : split) {
+            cars.add(new Car(s.trim()));
+        }
```
```suggestion
        List<Car> cars = new ArrayList<>()
        for (String s : split) {
            cars.add(new Car(s.trim()));
        }
```
여기서 ArrayList를 직접 생성해 구현하면 된다.

javajigi reviewed (changes requested):
피드백 반영한 부분에 대해 피드백할 부분이 있어 피드백 남겼어요.
앞 단계에서 피드백 반영하지 않은 부분 피드백 반영 후 다시 리뷰 요청 주세요. 👍

h2jinee committed:
feat: 자동차 생성 및 우승자 찾기 기능 추가

- CarFactory 클래스를 생성하여 자동차 목록을 생성하는 기능 추가
- RacingGame 클래스에서 자동차 목록을 관리하도록 수정
- OutputView 클래스에서 우승자 출력 방식을 개선
- 중복된 자동차 이름을 검증하는 로직 추가

h2jinee committed:
feat: 자동차 이름 제한 및 중복 검사 기능 추가

- CarFactory 클래스에 자동차 이름 최대 길이 제약 추가
- CarFactoryTest 클래스를 새로 생성하여 자동차 이름 중복 및 유효성 검사를 위한 테스트 추가
- CarTest 클래스에서 자동차의 초기 위치와 이동 관련 테스트 추가
- RacingGameTest 클래스 이름 변경

h2jinee commented:
@javajigi  포비님 남겨주신 피드백보고 수정했습니다.. 그런데 3단계에서 구현했던 RacingMain 클래스가 이제는 작동을 안 하는데요!! 어떻게 접근을 해야할지 잘 모르겠습니다. 🥹

javajigi commented on src/main/java/racinggame/domain/CarFactory.java:
```diff
@@ -0,0 +1,44 @@
+        return carNames;
+    }
+
+    private static void validateDuplicateNames(List<String> carNames) {
```
👍

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -41,14 +33,34 @@ private int generateRandomNumber() {
+        int maxPosition = 0;
+
+        for (Car car : cars) {
+            maxPosition = Math.max(maxPosition, car.getPosition());
```
car의 값을 꺼내지말고 객체에 메시지를 보내는 방식으로 접근해 볼 것을 추천

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -41,14 +33,34 @@ private int generateRandomNumber() {
+    }
+
+    private void addWinners(Car car, int maxPosition, List<String> winners) {
+        if (car.getPosition() == maxPosition) {
```
car의 값을 꺼내지말고 객체에 메시지를 보내는 방식으로 접근해 볼 것을 추천

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,62 @@
+    public static final int MAX_NAME_LENGTH = 5;
+
+    private int position;
+    private final String name;
```
'규칙 3: 모든 원시값과 문자열을 포장한다.'을 적용해 클래스를 분리해 보는 연습을 해볼 것을 추천
이 2개의 값 이외에도 자동차 경주 게임 미션에서 원시값과 문자열을 찾아 이 원칙을 적용해 본다.

javajigi commented on src/test/java/racinggame/domain/CarFactoryTest.java:
```diff
@@ -0,0 +1,50 @@
+        assertThat(cars).hasSize(3);
+        assertThat(cars.get(0).getName()).isEqualTo("pobi");
+        assertThat(cars.get(1).getName()).isEqualTo("crong");
+        assertThat(cars.get(2).getName()).isEqualTo("honux");
```
```suggestion
        List<Car> cars = createCars(names);

        assertThat(cars).hasSize(3);
        assertThat(cars).cotains(new Car("pobi"), new Car("crong"), new Car('honux"));
```
위와 같이 객체의 값을 꺼내 비교하기보다 객체 단위로 비교하는 방식으로 접근할 것을 추천

javajigi commented on src/test/java/racinggame/domain/WinnerTest.java:
```diff
@@ -0,0 +1,28 @@
+
+        pobi.move(5);
+        crong.move(5);
+        honux.move(3);
```
```suggestion
        Car pobi = new Car("pobi", 5);
        Car crong = new Car("crong", 5);
        Car honux = new Car("honux", 3);
```
위와 같이 position 값을 지정할 수 있는 생성자를 추가해 보면 어떨까?

javajigi commented on src/main/java/racinggame/RacingMain.java:
```diff
@@ -1,19 +1,22 @@
+
 
-public class RacingGame {
+public class RacingMain {
```
3단계에서 구현한 이 main() 메서드를 4단계에 맞도록 수정하는 것이 맞음.
RacingWinnerMain 내용을 이곳에 구현하면 됨
즉, 3단계에서 4단계로 요구사항이 추가되었으니 3단계와 같은 구현은 더 이상 실행되지 않아도 됨

javajigi reviewed (changes requested):
피드백 반영하느라 수고했어요. 👍 
지난 번 피드백에서 다른 부분 연습을 위해 추가하지 않았는데요.
객체 분리 연습을 위해 "규칙 3: 모든 원시값과 문자열을 포장한다." 적용에 대한 피드백 남겼으니 한번 도전해 보시죠.
RacingMain과 관련한 피드백 남겼으니 확인해 보세요.

h2jinee committed:
refactor: Car 클래스와 관련 클래스 구조 개선

- Car 클래스의 position과 name 필드를 CarPosition과 CarName 클래스로 변경하여 책임 분리
- Car 클래스의 move 메서드와 관련된 로직을 CarPosition으로 이동
- CarName 클래스를 추가하여 자동차 이름 검증 로직을 분리
- RacingGame 클래스에서 최대 위치 비교 로직을 Car 클래스의 메서드로 이동

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,70 @@
+    public Car(String name, int position) {
+        this(name);
+        this.position = new CarPosition(position);
+    }
```
```suggestion
    public Car(String name) {
        this(new CarName(name), new CarPosition(0));
    }

    public Car(String name, int position) {
        this(new CarName(name), new CarPosition(position));
    }
    
    public Car(CarName name, CarPosition position) {
        this.name = name;
        this.position = position;
    }
```
- 객체의 필드 값과 같은 타입의 주 생성자를 추가 
- 필드 값을 할당하는 작업은 주 생성자에서만
- 부 생성자는 주 생성자를 통해서만 필드 값을 초기화

생성자는 위와 같은 기준으로 구현할 것을 추천

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public boolean isWinnerWith(int targetPosition) {
+        return position.isGreaterThanOrEqual(targetPosition);
```
👍

javajigi commented on src/main/java/racinggame/domain/CarName.java:
```diff
@@ -0,0 +1,45 @@
+    private static final int MAX_NAME_LENGTH = 5;
+    private final String name;
+
+    public CarName(String name) {
```
👍

javajigi commented on src/main/java/racinggame/domain/CarPosition.java:
```diff
@@ -0,0 +1,53 @@
+        this(MIN_POSITION);
+    }
+
+    public CarPosition(int position) {
```
음수 값을 허용하지 않는다는 유효성 체크도 추가하면 어떨까?

javajigi commented on src/main/java/racinggame/domain/CarPosition.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public CarPosition move(int position) {
+        if (position >= MOVE_THRESHOLD) {
```
```suggestion
    public CarPosition move(int randomValue) {
        if (randomValue >= MOVE_THRESHOLD) {
```

CarPosition에 move까지 메시지를 보내는 것은 생각해 보지 않았는데 이 구현도 나름 일리가 있음.

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,66 @@
+        return maxPosition;
+    }
+
+    public List<String> findWinners() {
```
반환 값도 Winners와 같은 일급 콜렉션으로 구현해 보면 어떨까?

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,66 @@
+
+        for (Car car : cars) {
+            addWinners(car, maxPosition, winners);
+        }
```
인덴트가 2인 경우를 줄이려다보니 무리하게 메서드를 분리한 느낌이 든다.
stream api를 통해 구현해 보면 어떨까?

javajigi commented on src/test/java/racinggame/domain/CarFactoryTest.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Test
+    @DisplayName("5자를 초과하는 이름이 포함된 경우 IllegalArgumentException 예외를 발생시킨다")
+    void throwException_WhenNameExceedsFiveCharacters() {
```
이 테스트는 CarName 객체에서 유효성 체크를 하고 있다면 굳이 테스트하지 않아도 되지 않을까?

javajigi commented on src/main/java/racinggame/RacingMain.java:
```diff
@@ -0,0 +1,24 @@
+        printResults(count, racingGame);
+        printWinners(racingGame);
+
+        scanner.close();
```
MVC 패턴 구조로 구현한 샘플 코드 공유해 봄

```
public class RacingGameApplication {
    public static void main(String[] args) {
        String carNames = InputView.getCarNames();
        int tryNo = InputView.getTryNo();

        RacingGame racingGame = new RacingGame(carNames, tryNo);
        while(racingGame.isEnd()) {
            racingGame.race();
            ResultView.printResult(racingGame.getCars());
        }
        ResultView.printWinners(racingGame.findWinners());
    }
}
```

위 코드에서 힌트를 얻어 본인의 코드에 개선할 부분이 있는지 검토해 보고 리팩토링할 부분이 있다면 리팩토링에 도전해 보면 어떨까?

javajigi reviewed (changes requested):
원시 값, 문자열 포장 잘 했네요. 👍 
포장하면서 주 생성자/부 생성자 구현에 대해 습관화 했으면 하는 부분 피드백 남겼고요.
MVC 기반으로 개발한다고 가정했을 때 샘플 예제도 남겨봤어요.
시도 횟수의 경우도 원시 값인 만큼 원시 값을 포장해 보면 어떨까요?

바로 merge하려다 이번 미션이라도 확실히 연습하고 다음 미션으로 넘어가는 것이 더 좋겠다 생각해서요.
리뷰 요청이 반복되어 피곤하겠지만 조급함을 버리고 하나씩 익혀 간다는 마음으로 도전해 보면 좋겠습니다.

h2jinee committed:
feat: 자동차 경주 기능 개선 및 우승자 출력 방식 변경

- RacingMain 클래스에서 Scanner 사용을 제거하고, getCarNames 및 getTryCount 메서드를 통해 입력을 받도록 수정
- RacingGame 클래스의 생성자를 변경하여 자동차 이름과 시도 횟수를 직접 받아 처리하도록 개선
- 우승자 출력 방식을 변경하여 Winners 클래스를 통해 포맷팅된 결과를 출력하도록 수정
- CarPosition 클래스에서 move 메서드의 매개변수를 randomValue로 변경하고, validatePosition 메서드를 추가하여 위치 유효성 검사를 수행

javajigi commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,73 @@
+        this.name = new CarName(name);
+        this.position = position;
+    }
+
```
```suggestion
    public Car(CarName name, CarPosition position) {
        this.name = name;
        this.position = position;
    }
```
주 생성자는 필드와 완전히 일치하는 타입의 생성자를 주 생성자라 하고 주 생성자는 필수이다.

javajigi commented on src/main/java/racinggame/domain/CarPosition.java:
```diff
@@ -0,0 +1,60 @@
+        return position;
+    }
+
+    public CarPosition move(int randomValue) {
```
`int randomValue`도 원시값인데 포장하면 의미가 있을까?

javajigi commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,68 @@
+public class RacingGame {
+
+    private final List<Car> cars;
+    private final int tryCount;
```
```suggestion
    private final TryCount tryCount;
```
이 값도 원시값인데 위와 같이 포장하면 의미가 있을까?

javajigi reviewed (approved):
지금까지 수 많은 피드백을 반영하면서 4단계 미션 마무리하느라 수고했어요. 👍 
고생 많았습니다.
피드백 몇개 남겼는데 5단계 진행할 때 함께 적용해 보세요.

5단계 학습은 다음과 같이 해볼 것을 제안해 봅니다.
- 3차 라이브 강의 영상을 보고 느끼는 점이 있으면 해당 부분에 대한 리팩터링
- 어제 공유한 호석님 https://github.com/next-step/java-racingcar/blob/5e7c0dad4eba2d39c826ab99b7954390515fe083/docs/05-racing-car-refactoring.md 같은 목표 설정과 체크리스트를 만들고 의식적인 연습 도전
